### ISSUE TITLE
Inconsistent DPI with multimonitor client

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#1202]
*Issue migrated from trac ticket # 1202*

**component:** core | **priority:** major | **resolution:** fixed | **keywords:** dpi monitors

#### 2016-05-20 07:52:57: psycho_zs created the issue

___
> Using Xpra 0.17.2 on current Debian Testing.
> 
> When connecting from single monitor client, DPI is set correctly.
> When connecting from multi monitor client, DPI on server is set as 34x96, can not change it with xrandr on server.
> If connecting with --dpi 96, server DPI is set to 85x85, which is less weird than it was on 0.17.1, but still wrong.
> 
> 
> Client xrandr setup:
> ```
> $ xrandr 
> Screen 0: minimum 8 x 8, current 5440 x 1080, maximum 32767 x 32767
> LVDS1 connected 1600x900+1920+180 (normal left inverted right x axis y axis) 310mm x 174mm
>    1600x900      60.03*+  40.02  
>    1440x900      59.89  
>    1368x768      60.00  
>    1360x768      59.80    59.96  
>    1152x864      60.00  
>    1280x720      60.00  
>    1024x768      60.00  
>    1024x576      60.00  
>    960x540       60.00  
>    800x600       60.32    56.25  
>    864x486       60.00  
>    800x450       60.00  
>    640x480       59.94  
>    720x405       60.00  
>    640x360       60.00  
> DP1 disconnected (normal left inverted right x axis y axis)
> DP2 disconnected (normal left inverted right x axis y axis)
> DP3 disconnected (normal left inverted right x axis y axis)
> HDMI1 disconnected (normal left inverted right x axis y axis)
> HDMI2 connected primary 1920x1080+0+0 (normal left inverted right x axis y axis) 477mm x 268mm
>    1920x1080     60.00*+
>    1280x1024     75.02    60.02  
>    1152x864      75.00  
>    1024x768      75.08    60.00  
>    800x600       75.00    60.32  
>    640x480       75.00    60.00  
>    720x400       70.08  
> HDMI3 connected 1920x1080+3520+0 (normal left inverted right x axis y axis) 477mm x 268mm
>    1920x1080     60.00*+
>    1280x1024     75.02    60.02  
>    1152x864      75.00  
>    1024x768      75.08    60.00  
>    800x600       75.00    60.32  
>    640x480       75.00    60.00  
>    720x400       70.08  
> VGA1 disconnected (normal left inverted right x axis y axis)
> VIRTUAL1 disconnected (normal left inverted right x axis y axis)
> 
> $ xdpyinfo | grep inch
>   resolution:    96x96 dots per inch
> ```
> 
> Xpra server xrandr setup (when connected with --dpi 96)
> ```
> $ xrandr 
> xrandr: Failed to get size of gamma for output default
> Screen 0: minimum 320 x 175, current 3864 x 1050, maximum 8192 x 4096
> default connected 3864x1050+0+0 0mm x 0mm
>    8192x4096      5.00  
>    5120x3200     10.00  
>    3840x2880     10.00  
>    3840x2560     10.00  
>    3840x2048     10.00  
>    3840x2160     10.00  
>    2048x2048     10.00  
>    2560x1600     10.00  
>    1920x1440     20.00  
>    1920x1200     10.00  
>    1920x1080     10.00  
>    1600x1200     10.00  
>    1680x1050     10.00  
>    1600x900      20.00  
>    1400x1050     85.00    75.00    70.00    60.00  
>    1440x900      20.00  
>    1280x1024     20.00  
>    1366x768      60.00  
>    1280x800      20.00  
>    1024x768      25.00    20.00    87.00  
>    1024x600      25.00  
>    800x600       25.00  
>    320x200       85.00  
>    6400x4096      5.00  
>    4720x3840     10.00  
>    6400x2560      5.00  
>    5120x2880     10.00  
>    6400x2160     10.00  
>    5280x2560      5.00  
>    5120x2560      5.00  
>    5760x2160     10.00  
>    4400x2560     10.00  
>    5280x1920      5.00  
>    4096x2304     20.00  
>    4096x2160     10.00  
>    3360x2560     10.00  
>    5680x1440     15.00  
>    3120x2560     10.00  
>    3072x2560     10.00  
>    3000x2560     10.00  
>    4000x1660     20.00  
>    5496x1200     25.00  
>    4480x1440     10.00  
>    5280x1200     25.00  
>    4240x1440     10.00  
>    4160x1440     10.00  
>    4800x1200     10.00  
>    4000x1440     20.00  
>    3200x1800     30.00  
>    2560x2240     20.00  
>    5280x1080     25.00  
>    3904x1440     10.00  
>    3008x1692     20.00  
>    3840x1200     20.00  
>    2728x1680     25.00  
>    3600x1200     20.00  
>    3520x1196     20.00  
>    3840x1080     20.00  
>    3864x1050     20.00* 
>    3600x1080     20.00  
>    2560x1440     10.00  
>    2560x1400     50.00  
>    3288x1080     10.00  
>    3360x1050     60.00  
>    3200x1080     50.00  
>    3120x1050     60.00  
>    2048x1536     20.00  
>    1728x1520     50.00  
>    1856x1392     75.00    60.00  
>    1792x1344     75.00    60.00  
>    2048x1152     60.00    50.00  
>    1400x900      60.00  
>    1280x960      85.00    60.00  
>    1360x768      20.00  
>    800x1280      20.00  
>    1152x864     100.00    85.00    75.00    70.00    60.00  
>    800x1242      20.00  
>    1280x768      20.00  
>    1280x762      20.00  
>    1280x720      50.00    25.00  
>    720x1280      25.00  
>    720x1242      25.00  
>    1280x682      25.00  
>    768x1024      25.00    20.00  
>    768x986       25.00  
>    1024x730      25.00  
>    960x720       85.00    75.00    60.00  
>    1024x640      50.00  
>    928x696       75.00    60.00  
>    600x1024      25.00  
>    896x672       75.00    60.00  
>    600x986       25.00  
>    1024x576      60.00  
>    1024x562      25.00  
>    832x624       75.00  
>    960x540       60.00  
>    536x960       25.00  
>    960x536       25.00  
>    536x922       25.00  
>    600x800       25.00  
>    960x498       25.00  
>    600x762       25.00  
>    800x562       25.00  
>    840x525       85.00    75.00    70.00    60.00  
>    864x486       60.00  
>    480x854       25.00  
>    848x480       25.00  
>    480x810       25.00  
>    480x800       25.00  
>    800x480       25.00  
>    848x442       25.00  
>    700x525       85.00    75.00    70.00    60.00  
>    480x762       25.00  
>    640x512       85.00    75.00    60.00  
>    720x450       60.00  
>    640x480       85.00    75.00    73.00    60.00  
>    720x405       60.00  
>    720x400       85.00  
>    680x384       60.00  
>    640x400       85.00  
>    576x432      100.00    85.00    75.00    70.00    60.00  
>    640x360       49.00  
>    640x350       85.00  
>    512x384       87.00    85.00    75.00    70.00    60.00  
>    416x312       75.00  
>    400x300       85.00    75.00    72.00    60.00    56.00  
>    320x240       85.00    75.00    73.00    60.00  
>    360x200       85.00  
>    320x175       85.00 
> 
> $ xdpyinfo | grep inch
>   resolution:    85x85 dots per inch
> 
> # when connected without --dpi argument:
> $ xdpyinfo | grep inch
>   resolution:    34x96 dots per inch
> 
> ```
>

**Comment by totaam** [comment]
The desktop scaling is enabled automatically on large client screens to save bandwidth and it should preserve the same DPI when shown on the client, but only after adjusting it for the client side scaling: the virtual screen on the server is calculated to cover the exact same area (physical dimensions), just with fewer pixels.
For example: a scaling value of 200% (applies to both x and y dimensions) means that your virtual screen will have only 25% of the number of pixels found on the client's screen, and the DPI will also be 25% of that value so that applications should look the same, just with fewer pixels. (and more importantly you should be able to get 4 times more frames per second at the same quality setting)

But in order to get the hardware DPI (the one you get with xdpyinfo) to match your client's DPI, you need to use a patched dummy driver which we do not provide for debian systems, only for centos and fedora. See [DPI](../wiki/DPI) for more information.
Without the patched dummy driver, only the xrdb value will be set and some applications do not honour it.
The only workaround is to edit `/etc/xpra/xorg.conf` and change the physical screen  size, and / or the xvfb line in `/etc/xpra/xpra.conf` to use a specific DPI.

I will probably close this ticket as wontfix.

**Comment by totaam** [comment]
> The default software looks like it should have the patches.

No, you have the correct version for running xdummy, the DPI patches go on top: [/browser/xpra/tags/v0.17.x/rpmbuild] (0002, 0003 and 0004)/

> The page says that Ubuntu needs the user to be added to tty, video, and dialout groups. Is that needed for Debian as well?

No.

**Comment by totaam** [comment]
Since we will not provide patched dummy drivers for Debian, [r12658](../commit/0458106592e82c6569ca6acaca96a6fd74e6c856) makes `-dpi 96` the default on Debian systems.
 
Note: you should not be calling this "sync DPI" as this does no such thing, it only hardcodes a specific value and it will not adapt to your client's DPI settings.
As per the [DPI](../wiki/DPI), this will only work for applications that rely on the virtualized hardware DPI, and has no effect on the other settings - but the other settings should be synchronized with the client values properly anyway.

**Comment by totaam** [comment]
Please post the client and server output with `xpra -d screen ...`.
As well as all your config files and exact command lines.

**Comment by totaam** [comment]
There another occurrence of the bug #1193, causing the DPI to get doubled when used in the config file or command line.
You didn't attach the config file, but from what I am seeing, you must have set dpi=96 in there.

> So, the client seem to take physical monitors' info into account somehow. 

On X11, the xpra client uses the first valid value it finds from:
* the `--dpi=` switch (or its equivalent config file entry)
* the xrdb dpi entries: `Xft.dpi`, `Xft/DPI`, `gnome.Xft/DPI`
* the physical dimensions reported by the X11 server

> This data is just garbage and has no real application anywhere, it should be ignored.

Unfortunately, that is just not the case: the whole point the dummy driver patches you applied is to emulate this behaviour since that is what a large number of applications use for calculating the DPI. And this is what we use when everything else has failed.

**Comment by totaam** [comment]
> I actually did attach both configs in previous message.

My bad, missed it. BTW, in the future, please use attachments to keep tickets tidy.

> I always set DPI 96 in for my X server, and I've never seen any application that ignored it and used 102 or 131 DPI directly from monitor data.

Many applications do this, Java for example, and you may find more on this bug tracker.

> Anyway, how can monitor's physical DPI have any meaning in multimonitor setup?

The average value is better than nothing.
We also capture the per-monitor DPI, which has a meaning and which we will be able to use with other display servers, just not with X11.

> No app or toolkit is able to redraw itself when window is moved between monitors, DPI is always the same for the whole X screen.

Correct for X11.

**Comment by totaam** [comment]
> 52x96, everything is tiny. Why?

Those numbers look wrong, please post the "-d screen" debug log with latest trunk which includes better debug logging messages.
You can also run this command to get the same results:
```shell
$ XPRA_SCREEN_DEBUG=1 python -c "from xpra.x11.gtk2 import gdk_display_source;from xpra.platform.xposix import gui;print(gui._get_randr_dpi())"
2016-05-25 15:43:56,150 xdpi=93, ydpi=95 - size-mm=1573x577, size=5760x2160
(93, 95)
```

> Could you please add --ignore-everything-and-set-this-dpi-everywhere-possible= cli option?

That's already the case with the client's "--dpi=XXX" switch as long as you remove all dpi command line arguments to Xdummy and Xvfb (which is the default) and apply the patches (for applications relying on "hardware" dpi).
We then set the virtualized physical dpi and all the xrdb settings to match that value.

**Comment by totaam** [comment]
#### 2016-05-25 23:13:31: psycho_zs commented


___
> Yes!
> With latest trunk and clean configs starting with `--desktop-scaling=off --dpi 96` works as expected.
> Setting `dpi # 96`  and `desktop-scaling=off` in xpra.conf also works!
> 
> Could you point me to any info on status of upstream adoption of dpi patch in dummy driver?
> 
> BTW, build scripts which generate xpra.conf add build dir prefix to "/etc/xpra/xorg.conf" string in xvfb line.

**Comment by totaam** [comment]
> Could you point me to any info on status of upstream adoption of dpi patch in dummy driver?

Upstream is not going to be interested on the DPI patch as it is, because it is quite hackish.
The other patches are being pushed though, some have landed already.

Some links:
* #163
* #56
* [RFC: dummy driver RandR improvements?](https://lists.x.org/archives/xorg-devel/2014-January/040038.html) (no reply)
* [dummy ConstantDPI option](https://lists.x.org/archives/xorg-devel/2014-April/041879.html) (no reply)
* [dummy driver obvious DacSpeed fix](https://lists.x.org/archives/xorg-devel/2015-September/047310.html) (already applied)
* [xf86-video-dummy: fix pointer limits when resizing](https://lists.x.org/archives/xorg-devel/2016-April/049307.html) (in progress)
* [xserver forces 96 DPI on randr-1.2-capable drivers, overriding correct autodetection](https://bugs.freedesktop.org/show_bug.cgi?id=23705)
* [Gecko doesn't apply system default scale - X11](https://bugzilla.mozilla.org/show_bug.cgi?id=712898)


---

> BTW, build scripts which generate xpra.conf add build dir prefix to "/etc/xpra/xorg.conf" string in xvfb line.

The xvfb line should always contain the path to the xorg.conf that was installed with the xpra.conf.
And as far as I can tell, it is the case.
If you run "./setup.py build" then the path in build will be relative to the current working directory.
You can also try: `./setup.py install --home=./install`, etc...
Doesn't look like a bug in trunk to me (though #977 was fixed quite recently).

**Comment by totaam** [comment]
> I've built debian package using infrastructure from the trunk..

Gotcha: fixed in [r12685](../commit/5921e91a8e2cb63d10f8daf3fbe13063d24f8530)

**Comment by totaam** [comment]
If someone finds the time: #1215.

**Comment by totaam** [comment]
More DPI issues and regressions: #2610.
